### ISSUE TITLE
Support auto export of entries based on search criteria

### ISSUE THREAD

**Issue by Engrammae** [keepassxreboot/keepassxc#8049]
## Summary
Sharing credentials with KeeShare depending on a shared tag regardless of the group they belong to. At best, removing a credential from the share if the credential is untagged.

## Examples
Tag all credentials you want to share with _Person a_ with the tag _shared-with-person-a_ and do the same to share credentials with _Person b_.

## Context
Imagine A and B are subsets of all credentials (Root) and there is at least one credential in root that is not in A but in B and vice versa. So in short: <img src="http://www.sciweavers.org/tex2img.php?eq=A%2CB%20%5Csubset%20Root%2C%20A%20%5Cnot%5Csubset%20B%2C%20B%20%5Cnot%5Csubset%20A&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=0" align="center" border="0" alt="A,B \subset Root, A \not\subset B, B \not\subset A" width="206" height="18" />
I want to share the credentials of A with person a and respectively B with person b, then there is no way to structure my groups that way. Sharing credentials specified by a tag would allow overlapping but not disjoint sets.

**Comment by droidmonkey** [comment]
We are not going to implement something like that  because it will conflict directly with improvements being made to keeshare to isolate shares to a specific group.

**Comment by Engrammae** [comment]
could you elaborate what you mean by isolating shares to a specific group?

**Comment by droidmonkey** [comment]
Right now keeshare operates on the entry level. If you move an entry outside a keeshare group...it's still shared. Also if you delete a bunch of shared entries then try to resync a keeshare database it will actually delete entries in the database as well. It's just not a very good system. You are welcome to tag entries however you like, but we won't be using tags to determine share state.

**Comment by Engrammae** [comment]
ok, will there be a way to see the share state of an entry in the original database? Just in order to keep track?

**Comment by droidmonkey** [comment]
Yes once this work is done it will be made obvious which entries and groups are shared.

**Comment by bmcdonnell** [comment]
Coming from KeePass, I was very excited to see that KeePassXC already includes the KeeShare functionality, and supports tags. But I was quite disheartened to see

> We are not going to implement something like that

How are users supposed to share different database subsets with 2 or more others?

Also, I don't think it makes sense to require users to organize their groups/folders according to whom they share with, or the subset(s) they export.

FWIW, the KeePass (_not_ XC) plugin [KeePassSubsetExport](https://github.com/lukeIam/KeePassSubsetExport) includes the feature to export a subset with a specified tag. (No import, though.)

**Comment by droidmonkey** [comment]
Exporting and keeshare are completely different things.

**Comment by bmcdonnell** [comment]
@droidmonkey Thanks for your response. It seems you only responded to my last paragraph, but the more pertinent points in my previous comment are these:

> [Without this feature,] how are users supposed to share different database subsets with 2 or more others?
>
> Also, I don't think it makes sense to require users to organize their groups/folders according to whom they share with, or the subset(s) they export.

Could you address those?

**Comment by droidmonkey** [comment]
> how are users supposed to share different database subsets with 2 or more others

This really isn't a use case of KeePass based programs. Keeshare itself is a bit of a hack to get working in the KeePass construct. Merging database entries gets really tricky in these scenarios. Deleted entries propagate to your main database when you didn't intend it to. It's very messy under the hood.

WHICH, is why I am moving keeshare to be strongly bound to a subgroup and not be at the entry level like it currently is.

If you want to share entries with a lot of people you should be looking for a cloud based program instead where there is a single source of truth and gated access controls. Instead of multiple disparate databases trying desperately to be in sync with one another.

**Comment by bmcdonnell** [comment]
@droidmonkey thanks for the additional info. I don't fully understand it yet.

BTW, on further reflection, I think having a share associated with a group is not necessarily a problem for this proposed (and as-yet rejected) feature, as long as it would recurse into subgroups. The top level group could be used to search all groups, and tags could be added to the KeeShare configuration as filters.

> Deleted entries propagate to your main database when you didn't intend it to.

I imagine one way to address this would be to alert with a list of which entries were deleted from the subdatabase, and allow them to select which to keep/delete from the main database.

More completely, at import/merge time, I imagine UI elements listing the changed entries, summarizing their changes, and allowing further inspection (regular opening) of the entries from both databases, before deciding which to accept. The summary of changes could include e.g. datetime of latest revision of each; whether only 1 or both changed since their latest common revision; and listing which fields changed (e.g. username, password, group, tags, attachments, etc.) - not how they changed.

I think that would be in keeping with what you call the current "entry level" binding of this feature, that you're moving away from? But IMO (and in my limited time thinking about it so far) that would make this *so* much more useful.

> If you want to share entries with a lot of people you should be looking for a cloud based program instead where there is a single source of truth and gated access controls. Instead of multiple disparate databases trying desperately to be in sync with one another.

Rather, my current desired use case is duplicating overlapping subsets of a main database for a couple different devices, not necessarily people. The subsets generally are orthogonal to the group hierarchy. And I don't want to use any cloud hosting; files would be copied offline to sync.

**Comment by droidmonkey** [comment]
As I implement "keeshare 2.0" I'll think about more advanced export options, perhaps not a sync, which complicates things greatly, but an auto export based on criteria.

**Comment by bmcdonnell** [comment]
@droidmonkey thanks for your consideration. I'm interested to see what you come up with. :)
